### PR TITLE
clippy::complexity fixes

### PR DESCRIPTION
**Pull request by matthiaskrgr**
filter_map_identity
 needless_bool
 search_is_some
 unit_arg
 map_identity
 needless_question_mark
 derivable_impls

### CODE REVIEW

**Review by compiler-errors — APPROVED** [review]

File: compiler/rustc_const_eval/src/interpret/visitor.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub trait ValueVisitor<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>>: Sized {
     /// `read_discriminant` can be hooked for better error messages.
     #[inline(always)]
     fn read_discriminant(&mut self, v: &Self::V) -> InterpResult<'tcx, VariantIdx> {
-        Ok(self.ecx().read_discriminant(&v.to_op(self.ecx())?)?)
+        self.ecx().read_discriminant(&v.to_op(self.ecx())?)
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/visitor.rs:24]: I usually prefer consistently putting in the `?` and Ok-wrapping... but I guess it's futile to fight against clippy lints :/

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/interpret/visitor.rs:24]: I think in this case, the `())?)?)` is pretty hard to read

Comment (matthiaskrgr) [reply compiler/rustc_const_eval/src/interpret/visitor.rs:24]: Yeah, this is an `Ok(foo?)` which peels away one Ok on the inside just to wrap one around again on the outside :upside_down_face:

Comment (matthiaskrgr) [reply compiler/rustc_const_eval/src/interpret/visitor.rs:24]: (deleted, misread :sweat_smile:

**Pull request merged**